src/common/src/util/memcmp_encoding.rs

```diff
@@ -0,0 +1,558 @@
+            // TODO: need some test for this case (e.g. e2e test)
+            DataType::List { .. } => deserializer.skip_bytes()?,
+            DataType::Struct(t) => t
+                .fields
+                .iter()
+                .map(|field| {
+                    // use default null tags inside composite type
+                    calculate_encoded_size_inner(
+                        field,
+                        DEFAULT_NULL_TAG_NONE,
+                        DEFAULT_NULL_TAG_SOME,
+                        deserializer,
+                    )
+                })
+                .try_fold(0, |a, b| b.map(|b| a + b))?,
+            DataType::Jsonb => deserializer.skip_bytes()?,
+            DataType::Varchar => deserializer.skip_bytes()?,
+            DataType::Bytea => deserializer.skip_bytes()?,
+        };
+
+        // consume offset of fixed_type
+        if deserializer.position() == base_position + 1 {
+            // fixed type
+            deserializer.advance(len);
+        }
+    } else {
+        return Err(memcomparable::Error::InvalidTagEncoding(null_tag as _));
+    }
+
+    Ok(deserializer.position() - base_position)
+}
+
+pub fn encode_value(value: impl ToDatumRef, order: OrderType) -> Result<Vec<u8>> {
+    let mut serializer = memcomparable::Serializer::new(vec![]);
+    serialize_datum(value, order, &mut serializer)?;
+    Ok(serializer.into_inner())
+}
+
+pub fn decode_value(ty: &DataType, encoded_value: &[u8], order: OrderType) -> Result<Datum> {
+    let mut deserializer = memcomparable::Deserializer::new(encoded_value);
+    Ok(deserialize_datum(ty, order, &mut deserializer)?)
+}
+
+pub fn encode_array(array: &ArrayImpl, order: OrderType) -> Result<Vec<Vec<u8>>> {
+    let mut data = Vec::with_capacity(array.len());
+    for datum in array.iter() {
+        data.push(encode_value(datum, order)?);
+    }
+    Ok(data)
+}
+
+/// This function is used to accelerate the comparison of tuples. It takes datachunk and
+/// user-defined order as input, yield encoded binary string with order preserved for each tuple in
+/// the datachunk.
+pub fn encode_chunk(chunk: &DataChunk, column_orders: &[ColumnOrder]) -> Vec<Vec<u8>> {
+    let encoded_columns = column_orders
+        .iter()
+        .map(|o| encode_array(chunk.column_at(o.column_index).array_ref(), o.order_type).unwrap())
+        .collect_vec();
+
+    let mut encoded_chunk = vec![vec![]; chunk.capacity()];
+    for encoded_column in encoded_columns {
+        for (encoded_row, data) in encoded_chunk.iter_mut().zip_eq_fast(encoded_column) {
+            encoded_row.extend(data);
+        }
+    }
+
+    encoded_chunk
+}
+
+/// Encode a row into memcomparable format.
+pub fn encode_row(row: impl Row, column_orders: &[ColumnOrder]) -> Vec<u8> {
+    let mut encoded_row = vec![];
+    column_orders.iter().for_each(|o| {
+        encoded_row.extend(encode_value(row.datum_at(o.column_index), o.order_type).unwrap());
```

Comment:
It's not efficient to `extend` here. We'd better directly use the serializer.

Reply:
Same as above, it's old code just with filename renamed. May fix in later PR.

Reply:
Same as above, it's old code just with filename renamed. May fix in later PR.